packages/next/src/server/pipe-readable.ts

```diff
@@ -1,107 +1,115 @@
+import type { ServerResponse } from 'node:http'
+
+import './node-polyfill-web-streams'
+
 export function isAbortError(e: any): e is Error & { name: 'AbortError' } {
   return e?.name === 'AbortError'
 }
 
-/**
- * This is a minimal implementation of a Writable with just enough
- * functionality to handle stream cancellation.
- */
-export interface PipeTarget<R = any> {
-  /**
-   * Called when new data is read from readable source.
-   */
-  write: (chunk: R) => unknown
-
-  /**
-   * Always called once we read all data (if the writable isn't already
-   * destroyed by a client disconnect).
-   */
-  end: () => unknown
-
-  /**
-   * An optional method which is called after every write, to support
-   * immediately streaming in gzip responses.
-   */
-  flush?: () => unknown
-
-  /**
-   * The close event listener is necessary for us to detect an early client
-   * disconnect while we're attempting to read data. This must be done
-   * out-of-band so that we can cancel the readable (else we'd have to wait for
-   * the readable to produce more data before we could tell it to cancel).
-   */
-  on: (event: 'close', cb: () => void) => void
-
-  /**
-   * Allows us to cleanup our onClose listener.
-   */
-  off: (event: 'close', cb: () => void) => void
-
-  closed?: boolean
-}
+function createWriterFromResponse(
+  res: ServerResponse,
+  controller: AbortController
+): WritableStream<Uint8Array> {
+  let started = false
+  let finished = false
+  let aborted = false
```

Comment:
I'm pretty sure we can merge these into a single variable. `aborted` is always flipped with `finished`, and if we're `finished` we never need to abort.